oauth2/src/main/java/com/linecorp/armeria/internal/common/auth/oauth2/AbstractTokenOperationRequest.java

```diff
@@ -0,0 +1,113 @@
+@UnstableApi
+public abstract class AbstractTokenOperationRequest<T> extends AbstractOAuth2Request<T> {
+
+    /**
+     * A common abstraction for the requests implementing various Token operations request/response flows,
+     * such as Token Introspection flow
+     * (<a href="https://tools.ietf.org/html/rfc7662#section-2">[RFC7662], Section 2</a>),
+     * and Token Revocation flow (<a href="https://tools.ietf.org/html/rfc7009">[RFC7009]</a>).
+     *
+     * @param operationsEndpoint A {@link WebClient} to facilitate the Token Operations requests. Must
+     *                           correspond to the required Token Operations endpoint of the OAuth 2 system.
+     * @param operationsEndpointPath A URI path that corresponds to the token Operations endpoint of the
+     *                               OAuth 2 system.
+     * @param clientAuthorization Provides client authorization for the OAuth requests,
+     *                            as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected AbstractTokenOperationRequest(WebClient operationsEndpoint, String operationsEndpointPath,
+                                            @Nullable ClientAuthorization clientAuthorization) {
+        super(operationsEndpoint, operationsEndpointPath, clientAuthorization);
+    }
+
+    /**
+     * Makes a token operations (Introspection/Revocation) request using given {@code token} and
+     * optional {@code tokenType}.
+     * @param token A token this operation request applies to.
+     * @param tokenType A hint about the type of the token submitted for (Introspection/Revocation) operation.
+     *                  Either {@code access_token} or {@code refresh_token} as per
+     *                  <a href="https://tools.ietf.org/html/rfc7009#section-2.1">[RFC7009], Section 2.1</a>.
+     * @return A {@link CompletableFuture} carrying the target result.
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    public CompletableFuture<T> make(String token, @Nullable String tokenType) {
+        final QueryParamsBuilder requestFormBuilder = QueryParams.builder();
+
+        // populate request form data
+        // MANDATORY token
+        requestFormBuilder.add(TOKEN, Objects.requireNonNull(token, TOKEN));
```

Comment:
nit: How about checking for null of `token` before creating `QueryParamsBuilder`?

Reply:
changed